### ISSUE TITLE
Bug in extraction

### ISSUE THREAD

**Issue by Pratyush** [google-research/cryptanalytic-model-extraction#1]
Thanks for making the library open source! I was trying to run the examples from the README, but I got the following error when trying to extract:
```
$ python extract.py 10-15-15-1 42
START EXTRACTION ATTACK

Start running critical point search to find neurons on layer 0
At this iteration I have 0 critical points
Gathering 45 critical points
Start another sweep
Total intersections found 27
delta queries 0
Up to 1 of 45
Up to 2 of 45
Up to 3 of 45
Up to 4 of 45
Up to 5 of 45
Up to 6 of 45
Up to 7 of 45
Up to 8 of 45
Up to 9 of 45
Up to 10 of 45
Up to 11 of 45
Up to 12 of 45
Up to 13 of 45
Up to 14 of 45
Up to 15 of 45
Up to 16 of 45
Up to 17 of 45
Up to 18 of 45
Up to 19 of 45
Up to 20 of 45
Up to 21 of 45
Up to 22 of 45
Up to 23 of 45
Up to 24 of 45
Up to 25 of 45
Up to 26 of 45
Up to 27 of 45
Start another sweep
Total intersections found 27
delta queries 0
Up to 28 of 45
Up to 29 of 45
Up to 30 of 45
Up to 31 of 45
Up to 32 of 45
Up to 33 of 45
Up to 34 of 45
Up to 35 of 45
Up to 36 of 45
Up to 37 of 45
Up to 38 of 45
Up to 39 of 45
Up to 40 of 45
Up to 41 of 45
Up to 42 of 45
Up to 43 of 45
Up to 44 of 45
Up to 45 of 45
Query count after that search: 0
And now up to  45 critical points
After filtering duplicates we're down to  45 critical points
Start trying to do the graph solving
Go up to 1
0
Traceback (most recent call last):
  File "extract.py", line 252, in <module>
    run_full_attack()
  File "extract.py", line 53, in run_full_attack
    extracted_normal, extracted_bias, mask = layer_recovery.compute_layer_values(critical_points,
  File "/home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py", line 482, in compute_layer_values
    critical_groups, extracted_normals = graph_solve([x[0] for x in this_layer_critical_points],
  File "/home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py", line 72, in graph_solve
    batch_pairings = process_block(scaled_ratios, scaled_all_ratios)
  File "/home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py", line 40, in process_block
    pairings = jnp.sum(close, axis=2) >= max(MIN_SAME_SIZE,BLOCK_MULTIPLY_FACTOR*(jnp.log(ratios.shape[1])-2))
jax.traceback_util.FilteredStackTrace: jax.core.ConcretizationTypeError: Abstract tracer value encountered where concrete value is expected.

The problem arose with the `bool` function. 

While tracing the function process_block at /home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py:30, this value became a tracer due to JAX operations on these lines:

  operation w:float64[] = log v:float64[]
    from line /home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py:40 (process_block)

  operation y:float64[] = sub w:float64[] x:float64[]
    from line /home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py:40 (process_block)

  operation ba:float64[] = mul y:float64[] z:float64[]
    from line /home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py:40 (process_block)

  operation bc:bool[] = gt ba:float64[] bb:float64[]
    from line /home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py:40 (process_block)

See https://jax.readthedocs.io/en/latest/faq.html#abstract-tracer-value-encountered-where-concrete-value-is-expected-error for more information.

Encountered tracer value: Traced<ShapedArray(bool[])>with<DynamicJaxprTrace(level=0/1)>

The stack trace above excludes JAX-internal frames.
The following is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "extract.py", line 252, in <module>
    run_full_attack()
  File "extract.py", line 53, in run_full_attack
    extracted_normal, extracted_bias, mask = layer_recovery.compute_layer_values(critical_points,
  File "/home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py", line 482, in compute_layer_values
    critical_groups, extracted_normals = graph_solve([x[0] for x in this_layer_critical_points],
  File "/home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py", line 72, in graph_solve
    batch_pairings = process_block(scaled_ratios, scaled_all_ratios)
  File "/home/curunir/.local/lib/python3.8/site-packages/jax/traceback_util.py", line 139, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/curunir/.local/lib/python3.8/site-packages/jax/api.py", line 211, in f_jitted
    out = xla.xla_call(
  File "/home/curunir/.local/lib/python3.8/site-packages/jax/core.py", line 1162, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/curunir/.local/lib/python3.8/site-packages/jax/core.py", line 1153, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/home/curunir/.local/lib/python3.8/site-packages/jax/core.py", line 1165, in process
    return trace.process_call(self, fun, tracers, params)
  File "/home/curunir/.local/lib/python3.8/site-packages/jax/core.py", line 575, in process_call
    return primitive.impl(f, *tracers, **params)
  File "/home/curunir/.local/lib/python3.8/site-packages/jax/interpreters/xla.py", line 556, in _xla_call_impl
    compiled_fun = _xla_callable(fun, device, backend, name, donated_invars,
  File "/home/curunir/.local/lib/python3.8/site-packages/jax/linear_util.py", line 247, in memoized_fun
    ans = call(fun, *args)
  File "/home/curunir/.local/lib/python3.8/site-packages/jax/interpreters/xla.py", line 632, in _xla_callable
    jaxpr, out_avals, consts = pe.trace_to_jaxpr_final(fun, abstract_args)
  File "/home/curunir/.local/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 1038, in trace_to_jaxpr_final
    jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(fun, main, in_avals)
  File "/home/curunir/.local/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 1019, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers)
  File "/home/curunir/.local/lib/python3.8/site-packages/jax/linear_util.py", line 156, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py", line 40, in process_block
    pairings = jnp.sum(close, axis=2) >= max(MIN_SAME_SIZE,BLOCK_MULTIPLY_FACTOR*(jnp.log(ratios.shape[1])-2))
  File "/home/curunir/.local/lib/python3.8/site-packages/jax/core.py", line 504, in __bool__
    def __bool__(self): return self.aval._bool(self)
  File "/home/curunir/.local/lib/python3.8/site-packages/jax/core.py", line 870, in error
    raise_concretization_error(arg, fname_context)
  File "/home/curunir/.local/lib/python3.8/site-packages/jax/core.py", line 859, in raise_concretization_error
    raise ConcretizationTypeError(msg)
jax.core.ConcretizationTypeError: Abstract tracer value encountered where concrete value is expected.

The problem arose with the `bool` function. 

While tracing the function process_block at /home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py:30, this value became a tracer due to JAX operations on these lines:

  operation w:float64[] = log v:float64[]
    from line /home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py:40 (process_block)

  operation y:float64[] = sub w:float64[] x:float64[]
    from line /home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py:40 (process_block)

  operation ba:float64[] = mul y:float64[] z:float64[]
    from line /home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py:40 (process_block)

  operation bc:bool[] = gt ba:float64[] bb:float64[]
    from line /home/curunir/Development/cryptanalytic-model-extraction/src/layer_recovery.py:40 (process_block)

See https://jax.readthedocs.io/en/latest/faq.html#abstract-tracer-value-encountered-where-concrete-value-is-expected-error for more information.

Encountered tracer value: Traced<ShapedArray(bool[])>with<DynamicJaxprTrace(level=0/1)>

```

**Comment by carlini** [comment]
Huh. This is new. 

First, can you try to disable the jit (JAX_DISABLE_JIT=1) and see if this works around the issue?

If so, could you tell me what version of jax and jaxlib you're using and I'll try to debug.

**Comment by Pratyush** [comment]
Thanks for the prompt reply!

Running with `env JAX_DISABLE_JIT=1 python extract.py` works =)

`pip show jax` outputs
```
Name: jax
Version: 0.2.3
```

while `pip show jaxlib` outputs
```
Name: jaxlib
Version: 0.1.56
```

**Comment by carlini** [comment]
Huh. Alright. The big extraction might be quite slow without the jit enabled, but should still work if you're patient. I'll try to take a look at this to fix it properly after USENIX next week.

**Comment by Pratyush** [comment]
Heh, yeah it's pretty slow. Thanks for looking into this!

**Comment by carlini** [comment]
This should be fixed in 1093b62. Let me know how it goes.
